File: C/src/runge_kutta.c
// runge_kutta.c
// Implementation of the 4th order Runge-Kutta method

#include <stdio.h>
#include "runge_kutta.h"

void runge_kutta_4th_order(Func f, double y0, double t[], int n, double y[]) {
    y[0] = y0;

    for (int i = 1; i < n; i++) {
        double h = t[i] - t[i - 1];
        double k1 = h * f(t[i - 1], y[i - 1]);
        double k2 = h * f(t[i - 1] + h / 2, y[i - 1] + k1 / 2);
        double k3 = h * f(t[i - 1] + h / 2, y[i - 1] + k2 / 2);
        double k4 = h * f(t[i - 1] + h, y[i - 1] + k3);

        y[i] = y[i - 1] + (k1 + 2 * k2 + 2 * k3 + k4) / 6;
    }
}

File: R/src/runge_kutta.R
# runge_kutta.R
# Implementation of the 4th order Runge-Kutta method for solving ODEs

runge_kutta_4th_order <- function(dydt, y0, t) {
  # Arguments:
  # dydt - function representing the differential equation dy/dt = f(t, y)
  # y0 - initial condition
  # t - vector of time points
  
  n <- length(t)
  y <- numeric(n)
  y[1] <- y0
  
  for (i in 2:n) {
    h <- t[i] - t[i - 1]
    k1 <- h * dydt(t[i - 1], y[i - 1])
    k2 <- h * dydt(t[i - 1] + h / 2, y[i - 1] + k1 / 2)
    k3 <- h * dydt(t[i - 1] + h / 2, y[i - 1] + k2 / 2)
    k4 <- h * dydt(t[i - 1] + h, y[i - 1] + k3)
    
    y[i] <- y[i - 1] + (k1 + 2 * k2 + 2 * k3 + k4) / 6
  }
  
  return(y)
}
